src/clients/java/src/main/java/com/tigerbeetle/NativeClient.java

```diff
@@ -49,6 +49,12 @@ public void submit(final Request<?> request) throws ConcurrencyExceededException
         }
     }
 
+    public void release(final long packet) {
+        if (contextHandle == 0L)
```

Comment:
Hm, this feels suspicious. If we don't call release, then I think we leak the packet, and than the 

```zig
                    // We need to drain all free packets, to ensure that all
                    // inflight requests have finished.
                    while (self.packets_free.pop() != null) {
```

loop just loops forever? 

Or is it guaranteed that Java-side `close` won't return until all of the callbacks will have executed? But if that's the case, then we can assert here that context is not 0?

Reply:
Good catch. It's suspicious indeed ... I tried to write a comment, but I failed to explain it in simple terms.

We don't need to check `if (contextHandle == 0L)`  at the Java side for few reasons: 
- It will panic at Zig side if zero.
- It is always called from the JNI side, so it's unlike to be `0` without having segfaulted somewhere else before.
- Even if it's zero due to a `close` bug, we don't want to throw an exception there because it is called from the native thread, and we will have to store this exception to be popped up by the caller thread after completion. This kind of error is better represented with a `@panic`.

The truth is, we do  `if (contextHandle == 0L) return` for the sake of unit tests.

Some tests mock a callback simulator, passing arbitrary packet values which would segfault if the zig code tried to cast to `*tb_packet_t`, like this one:
https://github.com/tigerbeetledb/tigerbeetle/blob/c1cc4b53b81a2341069467ade49617f2431c3f7c/src/clients/java/src/test/java/com/tigerbeetle/AsyncRequestTest.java#L112-L115

Another solution would be a mocked client, but it seemed to be too much trouble to decouple a mock and still have `NativeClient.endRequest` properly asserted.

I agree that we need to `assert` if it happens outside unit tests.
Let me think of something else ...

Reply:
So, just to clarify, what would be the behavior on the java side if we do something like this?

```
var client = new Client();
CompletableFuture<CreateAccountResultBatch> accounts  = client.createAccountsAsync(...);
client.close();
var result = accounts.wait();
```

Reply:
`client.close()` waits until all requests are finished. So, it will block until `accounts` is processed. After that, `accounts.wait()` will return synchronously. Further requests will fail with `client closed`.